### ISSUE TITLE
Warning if Y extrapolation detected

### ISSUE THREAD

**Issue by dcornet** [GBeurier/nirs4all#13]
**Issue Summary:**
During model evaluation, we detected that some values of the target variable (Y) in the test set are outside the range observed in the training set. This means the model is being asked to extrapolate, which most machine learning models (especially tree-based models and neural networks) are not well equipped to handle.

**Why is this a problem?**
Extrapolation often leads to unreliable and unstable predictions. As a result, model performance metrics may be overly optimistic and not representative of real-world predictive ability.

**Recommendations to Address This:**
- Flag Out-of-Range Predictions: In reporting, flag any predictions that result from extrapolation.
- Use Adapted Scaling: Consider using a MinMaxScaler with an internal scaling range (e.g., [0.2, 0.8]) to allow a buffer for unseen values and reduce the risk of scaling artifacts.
- Constrain Model Output: Optionally, clip predictions to a plausible range or issue a warning when predictions exceed known boundaries.
- Quantify Uncertainty: Employ models or techniques that provide uncertainty estimates or prediction intervals, so extrapolated predictions can be interpreted with caution.
- Expand Training Data Diversity: If possible, retrain your model with data that better covers the range of real-world scenarios.
- Use Local Models: For critical cases, use local models (e.g., k-NN, lwPLS) or just-in-time approaches that refuse to predict when no similar training data is found.
- Regularize or Simplify Models: Apply stronger regularization to avoid wild extrapolations.
- Inform End Users: Always make clear to downstream users when extrapolation has occurred.

**Next Steps:**
Please review these warnings and consider adjusting your modeling and data splitting approach as suggested above to improve model reliability.

```
import numpy as np
import warnings
import matplotlib.pyplot as plt
import os

def warn_and_report_extrapolation(
    Y_train, Y_test, 
    buffer_range=(0.2, 0.8), 
    results_dir="results"
):
    os.makedirs(results_dir, exist_ok=True)
    
    y_min, y_max = np.min(Y_train), np.max(Y_train)
    test_min, test_max = np.min(Y_test), np.max(Y_test)
    print(f"Train Y range: [{y_min:.3f}, {y_max:.3f}]")
    print(f"Test Y range:  [{test_min:.3f}, {test_max:.3f}]")

    # Find indices and values of extrapolation in Y_test
    out_of_range_mask = (Y_test < y_min) | (Y_test > y_max)
    idx_extrap = np.where(out_of_range_mask)[0]
    values_extrap = Y_test[out_of_range_mask]
    if len(idx_extrap) > 0:
        print(f"Indices of extrapolation in test set: {list(idx_extrap)}")
        print(f"Values of extrapolation in test set: {list(values_extrap)}")
    else:
        print("No extrapolation detected in the test set.")

    # Save a histogram
    plt.figure(figsize=(8,4))
    plt.hist(Y_train, bins=20, alpha=0.6, label='Train Y', color='C0', edgecolor='black')
    plt.hist(Y_test, bins=20, alpha=0.6, label='Test Y', color='C1', edgecolor='black')
    if len(idx_extrap) > 0:
        plt.scatter(values_extrap, np.zeros_like(values_extrap), color='red', label='Extrapolated test Y', zorder=5)
    plt.axvline(y_min, color='C0', linestyle='dashed', linewidth=1)
    plt.axvline(y_max, color='C0', linestyle='dashed', linewidth=1)
    plt.xlabel("Y value")
    plt.ylabel("Count")
    plt.title("Distribution of Y in train and test sets")
    plt.legend()
    hist_path = os.path.join(results_dir, "y_train_test_histogram.png")
    plt.tight_layout()
    plt.savefig(hist_path)
    plt.close()
    print(f"Histogram comparing train and test Y saved at: {hist_path}")

    # Warning with recommendations
    if len(idx_extrap) > 0:
        warning_msg = (
            f"{len(idx_extrap)} test samples have Y values outside the training set range "
            f"(train Y: [{y_min:.2f}, {y_max:.2f}]; "
            f"test Y: [{test_min:.2f}, {test_max:.2f}]).\n"
            "⚠️ **Warning: The model will extrapolate for these test samples, which may result in unreliable predictions.**\n\n"
            "Recommendations to manage extrapolation:\n"
            "- Flag and report all extrapolated predictions in your output.\n"
            f"- Use a MinMaxScaler with a buffered range (e.g., {buffer_range}) to minimize scaling artifacts for future unseen values.\n"
            "- Consider constraining/clipping model outputs to physically plausible ranges and warn users when this occurs.\n"
            "- Where possible, use models that provide uncertainty estimates or prediction intervals, and highlight high-uncertainty predictions.\n"
            "- Expand your training data to cover a wider range if feasible, to reduce extrapolation risk in deployment.\n"
            "- For critical applications, use local modeling (e.g., k-NN, lwPLS, just-in-time models) that can refuse to predict when extrapolation is required.\n"
            "- Apply stronger regularization or choose simpler models to avoid unstable extrapolated predictions.\n"
            "- Inform all end-users or stakeholders whenever extrapolation occurs, so predictions are interpreted with appropriate caution.\n\n"
            f"See the saved histogram for a visual check: {hist_path}\n"
            f"Indices of extrapolation in test set: {list(idx_extrap)}\n"
            f"Values of extrapolation in test set: {list(values_extrap)}\n"
            "Please consider updating your data processing or modeling strategy accordingly."
        )
        warnings.warn(warning_msg)

    # Return useful info for further automated processing if needed
    return {
        "train_range": (y_min, y_max),
        "test_range": (test_min, test_max),
        "extrap_indices": idx_extrap,
        "extrap_values": values_extrap,
        "histogram_path": hist_path
    }
```
